### DISCUSSION TITLE
Possible memory leak?

### DISCUSSION THREAD

**Discussion by deku325** [Pryaxis/TShock#2879] [General]
I used Tshock to host a server to play with my friends and i noticed that when were playing for almost 1 hour my computer and the server were lagging, i opened the task manager to see if any process is causing lag and i saw that tshock was using almost 5 gb of ram, is this normal?, we are 6 players online in total, the only plugins installed were Crossplay and Vanillafier.

**Comment by drunderscore** [comment]
Would you be able to provide information such as your platform and logs on the day it reached l that usage?

**Reply by deku325** [reply]
My computer has a intel core i5 6400 and 12gb of ram with windows 10 so i downloaded tshock-5.1.3 for terraria 1.4.4.9 win-x64-release. 
There wasn't any errors in the logs.

**Reply by drunderscore** [reply]
Yes, depending on the issue, no errors in the log is expected.

I don't believe, of the few reports we have, witnessed a memory leak on Windows -- whether or not it is related to the others is still TBD... but unfortuantely with this little information it's unlikely to bring any fruit, without much further investigation (ex; heap analysis).

**Comment by sgkoishi** [comment]
Might be related: #2841, this is the third report of memory leak.
One (under Linux) said running [`dotnet gcdump collect`](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-gcdump#dotnet-gcdump-collect) (which triggers a full GC) temporary reduce the memory consumption, and the gcdump contains >500MB `System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs` and >1GB `byte[]`.
~~Tried to load the core dump but my VS is refusing it for unknown reason~~

**Reply by sgkoishi** [reply]
Would be quite helpful if there is a [`dump`](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump#dotnet-dump-collect) (but it might contain all sensitive information inside your server)

**Comment by sgkoishi** [comment]
Those `AwaitableSocketAsyncEventArgs` are tied to `SocketSendCallback`:

![image](https://user-images.githubusercontent.com/9637711/209074083-03592f31-d03a-4813-acd3-c5da45a60b15.png)

<hr>

There are around 4-5 million `BufferMemorySendOperation`s inside the thread pool queue and takes ~5GB.

![image](https://user-images.githubusercontent.com/9637711/209073711-f032765f-8143-499f-a73a-36808e5816dc.png)
